Reorganization of Expansions

z3dev opened the pull request:
**WORK IN PROGRESS. DO NOT MERGE.**

This pull request adds the expansion functions as a seperate module. The goals are:
- Convert to V2 math
- Standardize API; names, parameter passing, error checks, etc.
- Full test coverage

**Expansions:**

expand
- [X] code complete
- [X] tests complete, 89.47% coverage (87.5 / 90.0 / 84.44% coverage internally)
- [X] reveiw complete

Fixes #66 

offset
- [X] code complete
- [X] tests complete, 87.5% coverage (98.8 / 100 / 100% coverage internally)
- [X] reveiw complete

Fixes #75 
Fixes #131 
Fixes OpenJSCAD #351

Some small changes to math and geometry modulese are included.
- Added geom2.toOutlines()
- Added intersect() of two line segments to math/utils/
- Added area() under any number of ordered points to math/utils

z3dev committed:
fix(test) updated index to helpers

z3dev committed:
reorg(math) added intersect() of line segments to utils

z3dev committed:
reorg(geometry) added toOutlines function to geom2

z3dev committed:
reorg(operations) added expansions module with all functions and tests

z3dev committed:
reorg(math) moved area function to math utilities
adjusted dependent code

z3dev commented:
Examples of offset on paths.
- black path is the original
- green path is the offset path +3 delta
- red path is the offset path -3 delta

Segments of 0 (chamfer)
![offset_path_1](https://user-images.githubusercontent.com/8969165/63074419-2af95c80-bf68-11e9-9604-1ea445effe1a.png)
![offset_path_2](https://user-images.githubusercontent.com/8969165/63074420-2b91f300-bf68-11e9-9443-e4ffb7aa86b1.png)

Segments of 1 (edges meet at intersection)
![offset_path_3](https://user-images.githubusercontent.com/8969165/63074421-2b91f300-bf68-11e9-838d-5801cce097ce.png)

Segments of 32 (fillet with with rounded segments)
![offset_path_4](https://user-images.githubusercontent.com/8969165/63074422-2b91f300-bf68-11e9-9273-fe0d308a2bd1.png)

z3dev commented:
Examples of offset on 2D geometries.
- black line is the original
- green line is the offset +3 delta
- red line is the offset -3 delta

Chamfer of 0 (close corners)
![offset_geom2_4](https://user-images.githubusercontent.com/8969165/63074605-f0dc8a80-bf68-11e9-852c-36397599a19b.png)
![offset_geom2_5](https://user-images.githubusercontent.com/8969165/63074606-f0dc8a80-bf68-11e9-8b47-ac3f8295a42b.png)
![offset_geom2_6](https://user-images.githubusercontent.com/8969165/63074607-f0dc8a80-bf68-11e9-8440-6266fa26e484.png)

kaosat-dev commented:
This is gorgeous @z3dev ! Reviewing asap !

kaosat-dev commented on src/geometry/geom2/toOutlines.js:
```diff
@@ -0,0 +1,104 @@
+
+    // due to the logic of fromPoints()
+    // move the first point to the last
+    if (connectedVertexPoints.length > 0) {
```
I remember this kind of logic needed from V1, is it still applicable in V2 ?

z3dev replied:
I’ll review again.

z3dev replied:
Yes. This is still required as geom2.fromPoints() creates the first side from points[N-1] and points[0].

kaosat-dev commented on src/geometry/geom2/toOutlines.js:
```diff
@@ -0,0 +1,104 @@
+ * @param  {geom2} geometry
+ * @returns {Array} array of point sets, where each set is an array of points
+ */
+const toOutlines = (geometry) => {
```
I really like what this is for, lots of uses cases !

kaosat-dev commented on src/geometry/geom2/toOutlines.js:
```diff
@@ -0,0 +1,104 @@
+
+/** returns the outline(s) of the given 2D geometry as an array of points
+ * @param  {geom2} geometry
+ * @returns {Array} array of point sets, where each set is an array of points
```
this wording is confusing imho: (mostly the use of 'set' which is a data structure type) is this an array of arrays ? where each nested array is a 'side' ?

z3dev replied:
Updated in-line documentation.

kaosat-dev commented on src/geometry/geom2/toOutlines.js:
```diff
@@ -0,0 +1,104 @@
+        // more than one side starting at the same vertex
+        let bestangle = undefined
+        let startangle = vec2.angleDegrees(vec2.subtract(startside[1], startside[0]))
+        for (let sideindex = 0; sideindex < nextpossiblesides.length; sideindex++) {
```
wow

kaosat-dev commented on src/geometry/geom2/toOutlines.test.js:
```diff
@@ -0,0 +1,116 @@
+
+test('geom2: toOutlines() should return no outlines for empty geom2', (t) => {
+  let shp1 = create()
+  let exp1 = []
```
this is very good ! an empty array for an empty object makes manipulating this very nice!

kaosat-dev commented on src/geometry/geom2/toOutlines.test.js:
```diff
@@ -0,0 +1,116 @@
+  t.true(comparePoints(exp2, ret2))
+})
+
+test('geom2: toOutlines() should return outlines for edges that touch in geom2', (t) => {
```
this is very hard to visualize: what are 'touching' edges ?

z3dev replied:
Like a figure 8... where there's a point of intersection between two sides.

kaosat-dev commented on src/math/utils/area.js:
```diff
@@ -0,0 +1,14 @@
+/*
+ * Calculate the area under the given points
+ */
+const area = (points) => {
```
nice

kaosat-dev commented on src/math/utils/area.js:
```diff
@@ -0,0 +1,14 @@
+  return (area / 2.0)
+}
+
+module.exports = area
```
hmm thinking a bit more about this: this seems specifically for Vec2s (2d data) so I think it belongs there ?

kaosat-dev commented on src/math/utils/intersect.js:
```diff
@@ -0,0 +1,33 @@
+// @see http://paulbourke.net/geometry/pointlineplane/
+// Note: If the line segments do NOT intersect then undefined is returned
+// Return the intersection point of two line segments
+const intersect = (p1, p2, p3, p4) => {
```
usefull ! nice !

kaosat-dev commented on src/math/utils/intersect.js:
```diff
@@ -0,0 +1,33 @@
+// Calculate the intersect point of the two line segments (p1-p2 and p3-p4), end points included.
```
please switch to normal docstrings

kaosat-dev commented on src/math/utils/intersect.js:
```diff
@@ -0,0 +1,33 @@
+  return [x, y]
+}
+
+module.exports = intersect
```
hmm I am scratching my head over where this would fit best: it is also specific for 2D data, not sure what the right place would be ?

z3dev replied:
maybe the name should change? intersectLineSegments?

kaosat-dev commented on src/operations/expansions/expand.js:
```diff
@@ -0,0 +1,33 @@
+ * @example
+ * let newsphere = expand({delta: 2}, cube({center: [0,0,15], size: [20, 25, 5]}))
+ */
+const expand = (options, ...objects) => {
```
clean & awesome !

kaosat-dev commented on src/operations/expansions/expandGeom2.js:
```diff
@@ -0,0 +1,20 @@
+ * Expand the given geometry (geom2) using the given options (if any).
+ * @param {Object} options - options for expand
+ * @param {Number} [options.delta=1] - delta (+/-) of expansion
+ * @param {Integer} [options.segments=0] - number of segments when creating rounded corners, or zero for chamfer
```
hmm I both like this (using segments === 0 for chamfer) , if we had a non polygon based system, would you say that passing Infinity (an actual js thing) to this value would generate something with maximum resolution/ no polygons?

z3dev replied:
Yes. I think you are correct.
Maybe we would use 'fill' instead of segments? Where 'fill' means:
- 0 : close corners with a chamfer
- 1 : close corners with an edge
- N : close corners with a fillet (of course, we use count of segments now)

kaosat-dev commented on src/operations/expansions/expandPath2.js:
```diff
@@ -0,0 +1,71 @@
+ * @param {path2} geometry - the geometry to expand
+ * @returns {geom2} expanded geometry
+ */
+const expandPath2 = (options, geometry) => {
```
very awesome ! (despite my brain still breaking when thinking about what an expansion of a non closed path means ?)

z3dev replied:
I suggest we play with this once integrated into the viewer. It will be far easier to perceive, and changes may be needed.

z3dev replied:
@kaosat-dev do you remember this little piece of wonder?
```
    // Expand the path to a CAG
    // This traces the path with a circle with radius pathradius
  expandToCAG: function (pathradius, resolution)
```
So, expanding a path is not new. BUT, a whole lot more sensical.

z3dev replied:
Resolving. We can address this later if necessary.

kaosat-dev commented on src/operations/expansions/expandShell.js:
```diff
@@ -0,0 +1,215 @@
+
+const extrudePolygon = require('./extrudePolygon')
+
+const fnNumberSort = (a, b) => {
```
don't we have that in the utils ?

z3dev replied:
It will be when the PR for performance is merged. :)

z3dev replied:
Done

kaosat-dev commented on src/operations/expansions/extrudePolygon.js:
```diff
@@ -0,0 +1,32 @@
+
+// Extrude a polygon in the direction of the offsetvector.
+// Returns (geom3) a new geometry
+const extrudePolygon = (offsetvector, polygon1) => {
```
This is unrelated to extrusions I am guessing ? Should it not be 'offsetPolygon' ? (just asking, drawing a blank)

z3dev replied:
I’ll look again. It may be in the wrong place.

z3dev replied:
extrudePolygon() is being used by expandShell() for geom3 expansion. it's included here as I found no other use of this function in the library.

z3dev replied:
@kaosat-dev should I move extrudePolygon to extrusions?

kaosat-dev replied:
not sure to be honest : is it actual extrusion ? Ie like taking a flat shape and turning it 3D ?

z3dev replied:
Correct. It uses the polygon as the top and the bottom, and builds sides. The result is a new 3D geometry.

z3dev replied:
Let’s leave it here for now.

kaosat-dev commented on src/operations/expansions/inside.js:
```diff
@@ -0,0 +1,49 @@
+ * @param {Array} polygon - a list of points, where each point is an array with X and Y values
+ * @return {Integer} 1 if the point is inside, 0 if outside
+ */
+const isPointInside = (point, polygon) => {
```
these two are great, but I think they belong in a more generic place, for reuse (over the years I had a need for these a lot in designs

z3dev replied:
Sure. This is pretty generic. The polygon is defined as an ordered list of points

@kaosat-dev  So how about moving to utils?

kaosat-dev replied:
hmm I agree about generic, but 'utils' seems too generic: this will only work for 2D points, so I would
- put it under geom2 since we do not have 2d polygons as data structures? 
- rename the file to a more clear arePointsInsidePoly2.js

z3dev replied:
@kaosat-dev how about creating a poly2 data structure? The only uses now are:
- measureArea (now called area() is math utils)
- arePointsInside

It seems kind of a waste but really these functions would better align as part of this data structure. And who knows, later enhancements might require poly2.

kaosat-dev replied:
ok, agreed, not perfect, but coherent, and we can always change later !

z3dev replied:
Done. Create poly2 geometry with minimal functionality. Updated all expansion functions to use poly2 for mearsureArea and arePointsInside.

kaosat-dev commented on test/helpers/index.js:
```diff
@@ -1,5 +1,7 @@
 module.exports = {
+  comparePoints: require('./comparePoints'),
   comparePolygons: require('./comparePolygons'),
+  comparePolygonsAsPoints: require('./comparePolygonsAsPoints'),
```
I do not see this added as part of this PR ? or am I missing something

z3dev replied:
Just updating the index. The helpler function has been there for a long time.

kaosat-dev commented on src/operations/expansions/offset.js:
```diff
@@ -0,0 +1,29 @@
+ * @param {Object|Array} objects - object(s) to offset
+ * @return {Object|Array} the offset geometry(s)
+ */
+const offset = (options, ...objects) => {
```
ok I confess I am really confused: even going back & forth between files and your screenshots, I do not understand expand vs offset ?

z3dev replied:
Offset 
- path2 ; create offset of path with chamfers
- geom2 ; convert to outlines, offset each outline, combine points into geom2

Expand
- path2 ; create offsets on both sides, create ends, and convert to geom2
- geom2 ; same as offset
- geom3 ; old logic and very very slow

We could do more, but I stopped at this point.

kaosat-dev replied:
sorry for the slow reply on this one: if expand does the same as offset for geom2 it means to me that it is the same operand, so this seems very confusing as it is to be honest
it seems that the main issue here is expand/ offset for path2 since that is the only one that has different results.
I also have a hard time understanding (sorry!) why chamfer is a part of offset and not a different operation ?
I honestly like the results a LOT, but I really think more thought around the API is really worth it for this :)

z3dev replied:
‘Chamfer’ is my term for how to fill the corners, but I think this term makes sense. And we could use the term for rounded rectangles and rounded cubes as well. ‘Segments’ isn’t really the correct term.

https://en.m.wikipedia.org/wiki/Chamfer

The offset() function creates parallel line segments, but that leaves gaps on the corners. The ‘chamfer’ tells the routine how to fill those gaps.

Although, a chamfer() function may be possible I’m not sure it makes sense because the offset() function creates the gaps, and tracks the gaps. Its far more efficient to fill those gaps as part of offset().

z3dev replied:
RESET

kaosat-dev replied:
?? reset

z3dev replied:
I made some changes so let’s start again, with a new discussion.

kaosat-dev reviewed (changes requested):
Top notch work and very eager to try this out as a user @z3dev !
There are a few points I am unclear on however if you do not mind ?

z3dev committed:
docs(geom2) corrected docs for toOutlines

z3dev committed:
Merge branch 'V2' into reorg-expansions

z3dev committed:
fix(expansions) changed expandShell to require fnNumberSort from utils

z3dev commented:
![250E4424-9176-49C6-8BB0-D4B0EE10CC1B](https://user-images.githubusercontent.com/8969165/63643382-2839fc80-c70a-11e9-8e79-6970e7d9a926.jpeg)

(This is actually from a wood working website.)

z3dev commented:
For CAG V1, here's what expand() does to a shape with a hole (10x10 square with 5x5 hole). Notice that the hole is getting smaller, not larger.
```
    let obj1 = CAG.rectangle({radius: [10,10]});
    obj1 = obj1.subtract(CAG.rectangle({radius: [5,5]}));
    
    let obj2 = obj1.expand(1.0, 8);
```
![V1expand](https://user-images.githubusercontent.com/8969165/64657968-af100880-d470-11e9-9765-ed7a9d8a4163.png)

z3dev commented:
For V2, I think the same functionality should be offset. (See expand below)
```
  let obj1 = primitives.square({size: 10})
  obj1 = booleans.subtract(obj1, primitives.rectangle({size: [5, 5], center: [0,0]}))
  obj1.color = [1,1,0,1] // YELLOW

  let obj2 = expansions.offset({delta: 1}, obj1)
  obj2.color = [0,1,0,1] // GREEN
```
![V2offset](https://user-images.githubusercontent.com/8969165/64658238-bc79c280-d471-11e9-8158-69a4ea9ff4e8.png)

z3dev commented:
For V2, the expand makes the whole shape expand (enlarge), including the holes. Note, this is very simular to scale however scale cannot chamfer corners.
```
  let obj1 = primitives.square({size: 10})
  obj1 = booleans.subtract(obj1, primitives.rectangle({size: [5, 5], center: [0,0]}))
  obj1.color = [1,1,0,1] // YELLOW

  let obj2 = expansions.expand({delta: 1}, obj1)
  obj2.color = [0,1,0,1] // GREEN
```
![V2expand](https://user-images.githubusercontent.com/8969165/64658433-7bce7900-d472-11e9-9a06-5a8b1bc51dce.png)

z3dev committed:
fix(math) added proper documentation to intersect util function

z3dev committed:
fix(expansions) changed expand of geom2 to expand all outlines using the same offset
added test of expand with a complex geom2 (holes)

z3dev committed:
fix(expansions) added corners option for type of corners produced, no more overloaing the meaning of segments
adjusted all test suites

z3dev committed:
enhance(geometry) added basic poly2 geometry with minimum functionality
added test suites

z3dev committed:
fix(geometry) added poly2 to index

z3dev committed:
fix(poly2) added flip function for reversing rotation of points

z3dev committed:
fix(expansions) changed offsetFromPoints and offsetGeom2 to use poly2 functions

z3dev commented:
@kaosat-dev please review the changes to the API again.

Summary:
- added 'corners' option, which accepts 'edge', 'chamfer', 'round'
- corrected useage of 'segments' options, which is now only used for creating round corners

In addition, poly2 geometry has been created. And, all expansion functions now use poly2 for measureArea and arePointsInside.

kaosat-dev commented on src/geometry/poly2/arePointsInside.js:
```diff
@@ -0,0 +1,65 @@
+const measureArea = require('./measureArea')
```
great !

kaosat-dev commented on src/geometry/index.js:
```diff
@@ -2,5 +2,6 @@ module.exports = {
   geom2: require('./geom2'),
   geom3: require('./geom3'),
   path2: require('./path2'),
+  poly2: require('./poly2'),
```
groovy !

kaosat-dev commented on src/operations/expansions/expand.js:
```diff
@@ -9,9 +9,10 @@ const expandPath2 = require('./expandPath2')
- * @param {Integer} [options.segments] - number of segments when creating rounded corners, or zero for chamfer
- * @param {Object|Array} objects - the objects(s) to expand
+ * @param {Number} [options.delta=1] - delta (+/-) of expansion
+ * @param {String} [options.corners='round'] - type corner to create during of expansion; edge, chamfer, round
```
not perfect, but ok for now !

kaosat-dev commented on src/operations/expansions/expandGeom2.js:
```diff
@@ -6,24 +6,31 @@ const offsetFromPoints = require('./offsetFromPoints')
   const defaults = {
     delta: 1,
-    segments: 0
+    corners: 'round',
```
hmm I was gonna say 'all ok', but did we not say that 'edge' should be the default ? (as that has no approximations etc)

z3dev replied:
Yeah. I was going to make ‘edge’ the default, but V1 implementation is ‘round’.

I’m fine with edge but thought a short discussion would be best.

Thoughts?

z3dev replied:
Done. It's 'edge' now.

kaosat-dev reviewed (changes requested):
Groovy @z3dev  almost there !
Only main thing remaining is the default type of corner (see my specific comment): I though we discussed 'edge' being the more logical default ? or is that wrong ?

z3dev committed:
fix(geom2) removed unnecessary code from toOutlines

z3dev committed:
fix(expansions) changed the default for corners to 'edge'

z3dev commented:
@kaosat-dev please merge. :)
